File: go/post.go
package main

import (
        "fmt"
        "net/http"
        "net/url"
        "bytes"
        "encoding/base64"
        "encoding/hex"
        "crypto/md5"
        "log"
        "os"
        "strings"
        "io/ioutil"
        "time"
)

func main(){
    dirpath, _ := os.Getwd()
    path := dirpath + "/small.jpg"

    fmt.Println("file: ", path)

    // platform api`
    req_url := "http://deepi.sogou.com/api/sogouService"

    // service
    svc := "translateOcr"
    //svc := "basicOpenOcr"

    pid := "your pid"
    secret_key := "your key"
    
    file, _ := ioutil.ReadFile(path)
    pic_buf := base64.StdEncoding.EncodeToString([]byte(file))

    nouce := "43"

    len := strings.Count(pic_buf, "") - 1
    log.Println("len: ", len)
    if len > 1024 {
        len = 1024
    }

    // calculate sign
    rq := ToString(pid, svc, nouce, pic_buf[:len], secret_key)
    md5ctx := md5.New()
    md5ctx.Write([]byte(rq))
    cipherStr := md5ctx.Sum(nil)
    sign := hex.EncodeToString(cipherStr)

    log.Println("sign: ", sign)

    extParams := map[string] string {
        "pid" : pid,
        "service": svc,
        "salt" : nouce,
        "nouce" : nouce,
        "from" : "en",
        "lang" : "en",
        "to" : "zh-CHS",
        //"result_type": "text",
        //"direction_detect":"false",
        "sign" : sign,
        "image" : pic_buf,
    }


    start := GetTime()
    request, err := Post(req_url, extParams)
    if err != nil {
        log.Fatal(err)
    }

    client := &http.Client{}
    res, err := client.Do(request)
    if err != nil {
        log.Fatal(err)
    } else {
        body := &bytes.Buffer{}
        _, err := body.ReadFrom(res.Body)
        if err != nil {
            log.Fatal(err)
        }
        res.Body.Close()
        //fmt.Println("status: ", res.Status, ", header: ", res.Header)
        fmt.Println(body)
    }
    
    end := GetTime()
    log.Println("cost: ", end - start)
}


func Post(req_url string, params map[string] string) (*http.Request, error){

    v := url.Values{}

    for key, val := range params {
        v.Set(key, val)
    }

    s := v.Encode() 

    req, err := http.NewRequest("POST", req_url, strings.NewReader(s))

    if err != nil {
        return nil, err
    }

    req.Header.Add("Content-Type", "application/x-www-form-urlencoded")

    return req, err
}


func ToString( strs ... string) string {
    var buf bytes.Buffer
    for _, value := range strs {
        buf.WriteString(value)
    }

    return buf.String()
}


func GetTime() uint64 {
    return uint64(time.Now().UnixNano() / 1e6)
}

        

File: python/post.py
#!/usr/bin/env python

import urllib
import urllib2

import base64
import hashlib

import requests


service_url = "http://deepi.sogou.com/api/sogouService"


def Post(url, params):
    headers = {"Content-Type":"application/x-www-form-urlencoded"}
    req = urllib2.Request(url, urllib.urlencode(params), headers)
    response = urllib2.urlopen(req)
    return response.read()

def Post2(url, params):
    r = urllib.urlopen(url, urllib.urlencode(params))
    return r.read()
    
def Post3(url, params):
    r = requests.post(url, params)
    return r.text

def File2base64(file):
    f = open(file, 'rb')
    return base64.b64encode(f.read())

def CalSign(pid, service, salt, sign_image, key):
    m = hashlib.md5()
    data = pid + service + salt + sign_image + key
    print data
    m.update(data)
    return m.hexdigest()


if __name__ == '__main__': 

    img_file = 'ace.PNG'
    pid = 'your pid'
    key = 'your key'
    service = 'basicOpenOcr'
    lang = 'zh-CHS'
    salt = '201810'
    img_base64 = File2base64(img_file)
    sign = CalSign(pid, service, salt, img_base64[0:1024], key)
    print img_base64[0:1024]
    print sign

    param_data = {
        "pid":pid,
        "service":service,
        "lang":lang,
        "salt":salt,
        "image":img_base64,
        "sign":sign
        }
    
    #print param_data

    res = Post(service_url, param_data)
    print res
